netperf_tool/src/client_node.cpp

```diff
@@ -0,0 +1,148 @@
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+
+#include "utils.hpp"
+
+namespace netperf_tool
+{
+
+using namespace std::chrono_literals;
+
+ClientNode::ClientNode(
+  const rclcpp::NodeOptions & options,
+  const rclcpp::QoS & pub_qos,
+  size_t array_size,
+  std::chrono::nanoseconds target_pub_period,
+  std::chrono::nanoseconds server_timeout)
+: Node("client", "netperf_tool", options),
+  target_pub_period_{target_pub_period},
+  server_timeout_{server_timeout}
+{
+  pub_ = this->create_publisher<netperf_tool_interfaces::msg::Bytes>("test_topic", pub_qos);
+  // always keep a message with the data preallocated
+  msg_to_publish_.data.resize(array_size);
+  msg_to_publish_.id = 0;
+  rclcpp::SerializedMessage serialized_msg;
+  rclcpp::Serialization<netperf_tool_interfaces::msg::Bytes> serializer;
+  serializer.serialize_message(&msg_to_publish_, &serialized_msg);
+  collected_info_.number_of_messages_published = 0u;
+  collected_info_.serialized_message_size = serialized_msg.size();
+  client_ = this->create_client<netperf_tool_interfaces::srv::GetResults>("get_results");
+}
+
+bool
+ClientNode::wait_for_server()
+{
+  return client_->wait_for_service(server_timeout_);
+}
+
+void
+ClientNode::start_publishing()
+{
+  timer_ = this->create_wall_timer(
+    target_pub_period_,
+    std::bind(&ClientNode::pub_next_msg, this));
+}
+
+void
+ClientNode::stop_publishing()
+{
+  if (timer_) {
+    timer_->cancel();
+  }
+}
+
+void
+ClientNode::sync_with_server(rclcpp::Executor & exec)
+{
+  if (!pub_->wait_for_all_acked(server_timeout_)) {
+    RCLCPP_WARN(
+      this->get_logger(), "Some messages were not acknowledged by the netperf server ...");
+  }
+  if (!client_->service_is_ready()) {
+    RCLCPP_ERROR(
+      this->get_logger(), "Perf server is not available anymore, cannot get the results back");
+    return;
+  }
+  auto req = std::make_shared<netperf_tool_interfaces::srv::GetResults::Request>();
+  req->publisher_gid = this->get_stringified_pub_gid();
+  req->messages_total = collected_info_.number_of_messages_published;
+  auto future = client_->async_send_request(req);
+  exec.spin_until_future_complete(future);
+  statistics_ = future.get();
+}
+
+ClientResults
+ClientNode::extract_results()
+{
+  ClientResults results;
+  if (!statistics_) {
+    throw std::runtime_error{"you must first completely run the client before getting results"};
+  }
+  results.statistics = *statistics_;
+  statistics_.reset();
+  {
+    std::lock_guard guard{collected_info_mutex_};
+    results.collected_info = std::move(collected_info_);
+  }
+  return results;
+}
+
+std::string
+ClientNode::get_topic_name()
+{
+  return pub_->get_topic_name();
+}
+
+std::string
+ClientNode::get_stringified_pub_gid()
+{
+  return stringify_gid(pub_->get_gid());
+}
+
+void
+ClientNode::pub_next_msg()
+{
+  auto now = std::chrono::system_clock::now();
+  msg_to_publish_.timestamp = static_cast<size_t>(now.time_since_epoch().count());
```

Comment:
For example, if you used `builtin_interfaces/msg/Time` here you could just do:

```c++
msg_to_publish_.timestamp = this->now();
```

And if you wanted to ensure System time (rather than ROS Time which could be sim time), you can do that too:

```c++
auto now = rclcpp::Clock().now();  // default is system time, but could pass `RCL_SYSTEM_TIME`
msg_to_publish_.timestamp = now;
```

Reply:
Fixed in https://github.com/ros2/ros2cli/pull/720/commits/084b304892a29231a93a5a0f46fa8b9ad349b639.